File: HTCommander.Core/Utils/BinaryDataFile.cs
/*
Copyright 2026 Ylian Saint-Hilaire
Licensed under the Apache License, Version 2.0 (the "License");
http://www.apache.org/licenses/LICENSE-2.0
*/

using System;
using System.IO;
using System.Text;

namespace HTCommander
{
    public class BinaryDataFile
    {
        private readonly string _filePath;
        private FileStream _fileStream;
        private BinaryWriter _writer;
        private BinaryReader _reader;
        private bool _isOpened = false;

        // Constants for data types
        private const int DataTypeInt = 1;
        private const int DataTypeString = 2;
        private const int DataTypeByteArray = 3;

        public BinaryDataFile(string filePath)
        {
            _filePath = filePath;
        }

        /// <summary>
        /// Opens the binary file for reading and writing.
        /// </summary>
        public void Open()
        {
            if (!_isOpened)
            {
                _fileStream = new FileStream(_filePath, FileMode.OpenOrCreate, FileAccess.ReadWrite);
                _writer = new BinaryWriter(_fileStream);
                _reader = new BinaryReader(_fileStream);
                _isOpened = true;
            }
        }

        /// <summary>
        /// Closes the binary file and releases associated resources.
        /// </summary>
        public void Close()
        {
            if (_isOpened)
            {
                _writer?.Dispose();
                _reader?.Dispose();
                _fileStream?.Dispose();
                _writer = null;
                _reader = null;
                _fileStream = null;
                _isOpened = false;
            }
        }

        /// <summary>
        /// Sets the file stream's seek pointer to the beginning of the file.
        /// </summary>
        public void SeekToBegin()
        {
            if (_isOpened && _fileStream != null)
            {
                _fileStream.Seek(0, SeekOrigin.Begin);
            }
        }

        /// <summary>
        /// Sets the file stream's seek pointer to the end of the file.
        /// </summary>
        public void SeekToEnd()
        {
            if (_isOpened && _fileStream != null)
            {
                _fileStream.Seek(0, SeekOrigin.End);
            }
        }

        /// <summary>
        /// Appends a new record to the binary file. The file must be opened before calling this method.
        /// </summary>
        /// <param name="userType">The user-defined type of the data.</param>
        /// <param name="data">The data to append (can be an int, string, or byte array).</param>
        /// <exception cref="InvalidOperationException">Thrown if the file is not open.</exception>
        public void AppendRecord(int userType, object data)
        {
            if (data == null) return;
            if (!_isOpened || _writer == null)
            {
                throw new InvalidOperationException("The file must be opened before appending a record.");
            }

            int dataType;
            int dataLength = 0;
            byte[] dataBytes = null;

            if (data is int intValue)
            {
                dataType = DataTypeInt;
                dataLength = sizeof(int);
                dataBytes = BitConverter.GetBytes(intValue);
            }
            else if (data is string stringValue)
            {
                dataType = DataTypeString;
                dataBytes = Encoding.UTF8.GetBytes(stringValue);
                dataLength = dataBytes.Length;
            }
            else if (data is byte[] byteArrayValue)
            {
                dataType = DataTypeByteArray;
                dataBytes = byteArrayValue;
                dataLength = byteArrayValue.Length;
            }
            else
            {
                throw new ArgumentException("Unsupported data type. Must be int, string, or byte array.");
            }

            // Calculate total record size
            int recordSize = sizeof(int) + sizeof(int) + sizeof(int) + dataLength;

            // Write the record
            _writer.Write(recordSize);
            _writer.Write(userType);
            _writer.Write(dataType);
            _writer.Write(dataBytes);
        }

        /// <summary>
        /// Reads the next record from the currently opened file and advances the stream pointer.
        /// </summary>
        /// <returns>A tuple containing the user type and the data, or null if the end of the file is reached.</returns>
        /// <exception cref="InvalidOperationException">Thrown if the file is not open.</exception>
        public int ReadNextRecord(out object data)
        {
            data = null;
            if (!_isOpened || _reader == null || _fileStream == null)
            {
                throw new InvalidOperationException("The file must be opened before reading a record.");
            }

            if (_fileStream.Position >= _fileStream.Length)
            {
                return 0; // End of file reached
            }

            int recordSize = _reader.ReadInt32();
            int userType = _reader.ReadInt32();
            int dataType = _reader.ReadInt32();

            switch (dataType)
            {
                case DataTypeInt:
                    data = _reader.ReadInt32();
                    break;
                case DataTypeString:
                    int stringLength = recordSize - (sizeof(int) * 3);
                    byte[] stringBytes = _reader.ReadBytes(stringLength);
                    data = Encoding.UTF8.GetString(stringBytes);
                    break;
                case DataTypeByteArray:
                    int byteArrayLength = recordSize - (sizeof(int) * 3);
                    data = _reader.ReadBytes(byteArrayLength);
                    break;
                default:
                    // Skip the rest of the record if the data type is unknown
                    int bytesToSkip = recordSize - (sizeof(int) * 3);
                    _fileStream.Seek(bytesToSkip, SeekOrigin.Current);
                    throw new InvalidDataException($"Unknown data type: {dataType}");
            }

            return userType;
        }
    }
}

File: htcommander_flutter/lib/radio/binary_data_file.dart
/*
Copyright 2026 Ylian Saint-Hilaire
Licensed under the Apache License, Version 2.0 (the "License");
http://www.apache.org/licenses/LICENSE-2.0
*/

/// Structured binary record file I/O.
/// Port of HTCommander.Core/Utils/BinaryDataFile.cs
///
/// Records are stored as:
///   [recordSize:4] [userType:4] [dataType:4] [data:variable]
/// where dataType is 1=int, 2=string(UTF-8), 3=byte array.
library;

import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

class BinaryDataFile {
  final String _filePath;
  RandomAccessFile? _file;
  bool _isOpened = false;

  // Data type constants
  static const int _dataTypeInt = 1;
  static const int _dataTypeString = 2;
  static const int _dataTypeByteArray = 3;

  BinaryDataFile(this._filePath);

  /// Opens the binary file for reading and writing.
  void open() {
    if (!_isOpened) {
      final file = File(_filePath);
      _file = file.openSync(mode: FileMode.append);
      _isOpened = true;
    }
  }

  /// Closes the binary file and releases resources.
  void close() {
    if (_isOpened) {
      _file?.closeSync();
      _file = null;
      _isOpened = false;
    }
  }

  /// Sets the file position to the beginning.
  void seekToBegin() {
    if (_isOpened && _file != null) {
      _file!.setPositionSync(0);
    }
  }

  /// Sets the file position to the end.
  void seekToEnd() {
    if (_isOpened && _file != null) {
      _file!.setPositionSync(_file!.lengthSync());
    }
  }

  /// Appends a record. [data] can be [int], [String], or [Uint8List].
  void appendRecord(int userType, Object data) {
    if (!_isOpened || _file == null) {
      throw StateError('File must be opened before appending a record.');
    }

    int dataType;
    Uint8List dataBytes;

    if (data is int) {
      dataType = _dataTypeInt;
      dataBytes = Uint8List(4);
      ByteData.view(dataBytes.buffer).setInt32(0, data, Endian.little);
    } else if (data is String) {
      dataType = _dataTypeString;
      dataBytes = Uint8List.fromList(utf8.encode(data));
    } else if (data is Uint8List) {
      dataType = _dataTypeByteArray;
      dataBytes = data;
    } else {
      throw ArgumentError(
          'Unsupported data type. Must be int, String, or Uint8List.');
    }

    // recordSize = 3 ints (12 bytes header) + data length
    final recordSize = 12 + dataBytes.length;
    final header = Uint8List(12);
    final bd = ByteData.view(header.buffer);
    bd.setInt32(0, recordSize, Endian.little);
    bd.setInt32(4, userType, Endian.little);
    bd.setInt32(8, dataType, Endian.little);

    // Seek to end before writing
    _file!.setPositionSync(_file!.lengthSync());
    _file!.writeFromSync(header);
    _file!.writeFromSync(dataBytes);
  }

  /// Reads the next record from the file.
  /// Returns a record with [userType] and [data], or null at end of file.
  /// [data] will be [int], [String], or [Uint8List].
  ({int userType, Object data})? readNextRecord() {
    if (!_isOpened || _file == null) {
      throw StateError('File must be opened before reading a record.');
    }

    if (_file!.positionSync() >= _file!.lengthSync()) {
      return null; // End of file
    }

    final header = _file!.readSync(12);
    if (header.length < 12) return null;

    final bd = ByteData.view(Uint8List.fromList(header).buffer);
    final recordSize = bd.getInt32(0, Endian.little);
    final userType = bd.getInt32(4, Endian.little);
    final dataType = bd.getInt32(8, Endian.little);
    final dataLength = recordSize - 12;

    Object data;
    switch (dataType) {
      case _dataTypeInt:
        final bytes = _file!.readSync(4);
        data =
            ByteData.view(Uint8List.fromList(bytes).buffer)
                .getInt32(0, Endian.little);
        break;
      case _dataTypeString:
        final bytes = _file!.readSync(dataLength);
        data = utf8.decode(bytes);
        break;
      case _dataTypeByteArray:
        data = Uint8List.fromList(_file!.readSync(dataLength));
        break;
      default:
        // Skip unknown data type
        _file!.setPositionSync(_file!.positionSync() + dataLength);
        throw FormatException('Unknown data type: $dataType');
    }

    return (userType: userType, data: data);
  }
}
